# HG changeset patch
# Parent 85cac8a02c040b28587794fc806365fd08f11f0b
xen/x86: make __direct_remap_pfn_range()'s return value meaningful

From: Olaf Hering <ohering@novell.com>

This change fixes the xc_map_foreign_bulk interface, which would
otherwise cause SIGBUS when pages are gone because -ENOENT is not
returned as expected by the IOCTL_PRIVCMD_MMAPBATCH_V2 ioctl.

Signed-off-by: Jan Beulich <jbeulich@novell.com>
Ported-by and 
Signed-off-by: Andres Lagar-Cavilla <andres@lagarcavillai.org>

diff -r 85cac8a02c04 arch/x86/mm/ioremap-xen.c
--- a/arch/x86/mm/ioremap-xen.c
+++ b/arch/x86/mm/ioremap-xen.c
@@ -68,8 +68,8 @@ static int __direct_remap_pfn_range(stru
 						 direct_remap_area_pte_fn, &w);
 			if (rc)
 				goto out;
-			rc = -EFAULT;
-			if (HYPERVISOR_mmu_update(u, v - u, NULL, domid) < 0)
+			rc = HYPERVISOR_mmu_update(u, v - u, NULL, domid);
+			if (rc < 0)
 				goto out;
 			v = w = u;
 			start_address = address;
@@ -94,8 +94,8 @@ static int __direct_remap_pfn_range(stru
 					 direct_remap_area_pte_fn, &w);
 		if (rc)
 			goto out;
-		rc = -EFAULT;
-		if (unlikely(HYPERVISOR_mmu_update(u, v - u, NULL, domid) < 0))
+		rc = HYPERVISOR_mmu_update(u, v - u, NULL, domid);
+		if (unlikely(rc < 0))
 			goto out;
 	}
 
